doom/.config/doom/config.org
#+title: Emacs Config
#+author: Zachary D Meyer
#+OPTIONS: toc: 5
#+startup: showeverything

* Table of Contents :toc:
- [[#configel][Config.el]]
  - [[#opening-comments][Opening Comments]]
  - [[#org-directory][Org Directory]]
  - [[#org-directory-initel][Org Directory init.el]]
  - [[#themes][Themes]]
  - [[#org-mode][Org Mode]]
  - [[#display-line-numbers-type][Display Line Numbers Type]]
  - [[#development-environment][Development Environment]]
  - [[#syntax-highlighting-changes-for-files][Syntax Highlighting Changes For Files]]
  - [[#custom-keybinds][Custom Keybinds]]
  - [[#documentation-comments][Documentation Comments]]
  - [[#closing-comments][Closing Comments]]
- [[#packagesel][Packages.el]]
  - [[#opening-comments-1][Opening Comments]]
  - [[#programming][Programming]]
  - [[#doxygen][DOXYGEN]]
  - [[#org-mode-packages][ORG MODE PACKAGES]]
  - [[#paredit][Paredit]]
  - [[#terminal][Terminal]]

* Config.el

** Opening Comments
This was here when I first got it.

#+begin_src emacs-lisp :tangle config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
;; (setq user-full-name "John Doe"
;;       user-mail-address "john@doe.coM")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-symbol-font' -- for symbols
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:

#+end_src

** Org Directory
My org directory is just ~/org Use a function stolen online to get every file.

#+begin_src emacs-lisp :tangle config.el
;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!

(defun sa-find-org-file-recursively (&optional directory filext)
  "Return .org and .org_archive files recursively from DIRECTORY.
If FILEXT is provided, return files with extension FILEXT instead."
  (interactive "DDirectory: ")
  (let* (org-file-list
         (case-fold-search t)         ; filesystems are case sensitive
         (file-name-regex "^[^.#].*") ; exclude dot, autosave, and backupfiles
         (filext (or filext "org$\\\|org_archive"))
         (fileregex (format "%s\\.\\(%s$\\)" file-name-regex filext))
         (cur-dir-list (directory-files directory t file-name-regex)))
    ;; loop over directory listing
    (dolist (file-or-dir cur-dir-list org-file-list) ; returns org-file-list
      (cond
       ((file-regular-p file-or-dir)             ; regular files
        (if (string-match fileregex file-or-dir) ; org files
            (add-to-list 'org-file-list file-or-dir)))
       ((file-directory-p file-or-dir)
        (dolist (org-file (sa-find-org-file-recursively file-or-dir filext)
                          org-file-list) ; add files found to result
          (add-to-list 'org-file-list org-file)))))))

(setq org-directory "~/org/")
(setq org-roam-directory "~/org/")

(defun zdm-org-agenda-refresh-files ()
  (setq org-agenda-files (append (sa-find-org-file-recursively "~/org/"))))

(add-hook 'after-save-hook 'zdm-org-agenda-refresh-files)
#+end_src

** Org Directory init.el
Run ~/org/elisp/init.el upon reloading config.

#+begin_src emacs-lisp
(defvar zdm-org-elisp-directory-path
  (concat org-directory "elisp/")
  "Private org mode ELISP directory path")

(load-file (concat "~/org/" "elisp/init.el"))
#+end_src

** Themes
Keep the dark mode theme for now.

#+begin_src emacs-lisp :tangle config.el
;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'deeper-blue)
#+end_src

** Org Mode
This section contains my org mode tweaks.

*** Header Size Customization
Some variety in header sizes.

#+begin_src emacs-lisp :tangle config.el
;; Set the font sizes for the org levels.
(custom-set-faces
 '(org-level-1 ((t (:inherit outline-1 :height 1.6))))
 '(org-level-2 ((t (:inherit outline-2 :height 1.3))))
 '(org-level-3 ((t (:inherit outline-3 :height 1.21))))
 '(org-level-4 ((t (:inherit outline-4 :height 1.15))))
 '(org-level-5 ((t (:inherit outline-5 :height 1.09))))
)
#+end_src

*** Hide Emphasis Markers
#+begin_src emacs-lisp :tangle config.el
(setq org-hide-emphasis-markers t)
#+end_src

*** Enable Org-Bullets-Mode
This code will hook org-mode to enable org-bullets-mode.
#+begin_src emacs-lisp :tangle config.el
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
#+end_src

** Display Line Numbers Type
#+begin_src emacs-lisp :tangle config.el
;; This determines the style of line numbers in effect. If set to `nil', line
; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)

#+end_src

** Development Environment
This section outlines the development environment within Emacs.
Currently we only require dap-gdb-lldb.

Only C++ is considered for now.

#+begin_src emacs-lisp :tangle config.el
(require 'dap-gdb-lldb)
#+end_src

*** C++ Environment
#+begin_src emacs-lisp :tangle config.el
;;
;; C++ Development
;;
(with-eval-after-load 'lsp-clangd
  (require 'dap-cpptools)
  (require 'dap-lldb))

#+end_src

*** Copilot
#+begin_src emacs-lisp :tangle config.el
(use-package! copilot
  :hook (prog-mode . copilot-mode)
  :bind (:map copilot-completion-map
              ("<tab>" . 'copilot-accept-completion)
              ("TAB" . 'copilot-accept-completion)
              ("C-TAB" . 'copilot-accept-completion-by-word)
              ("C-<tab>" . 'copilot-accept-completion-by-word)
              ("C-n" . 'copilot-next-completion)
              ("C-p" . 'copilot-previous-completion))

  :config
  (add-to-list 'copilot-indentation-alist '(prog-mode 2))
  (add-to-list 'copilot-indentation-alist '(org-mode 2))
  (add-to-list 'copilot-indentation-alist '(text-mode 2))
  (add-to-list 'copilot-indentation-alist '(emacs-lisp-mode 2)))
#+end_src

** Syntax Highlighting Changes For Files
This section has configuration on specific file types that do not have the
correct syntax highlighting.

*** C++
This subsection contains all of the C++ files that need to be in C++ mode.
Currently, it's just the .inl and .tpp files.

#+begin_src emacs-lisp :tangle config.el
;;
;; C++ file syntax highlighting
;;
(add-to-list 'auto-mode-alist '(".*\.inl" . cpp-mode))
(add-to-list 'auto-mode-alist '(".*\.tpp" . cpp-mode))

(eval-after-load 'flycheck
  '(add-hook 'flycheck-mode-hook #'flycheck-clang-tidy-setup))
(setq flycheck-clang-tidy-executable "clang-tidy-19")
#+end_src


** Custom Keybinds
This section contains my custom keybindings for Doom Emacs.
Some of this stuff has been copied from DistroTube, but I'll make it my own.

*** Dired Keybinds
Here are the keybindings for the dired file manager.

|-------------+--------------------+------------------|
| Key Binding | Command            | Description      |
|-------------+--------------------+------------------|
| M-RET       | dired-display-file | Display File     |
| h           | dired-up-directory | Up one directory |
| l           | dired-open-file    | Open file        |
| m           | dired-mark         | Mark file        |
|-------------+--------------------+------------------|


#+begin_src emacs-lisp :tangle config.el
;;
;; Dired Keybindings
;;
(map! :leader
      (:prefix ("d" . "dired")
       :desc "Open dired" "d" #'dired
       :desc "Dired jump to current" "j" #'dired-jump)
      (:after dired
        (:map dired-mode-map
         :desc "Peep-dired image previews" "d p" #'peep-dired
         :desc "Dired view file" "d v" #'dired-view-file)))

(evil-define-key 'normal dired-mode-map
  (kbd "M-RET") 'dired-display-file
  (kbd "h") 'dired-up-directory
  (kbd "l") 'dired-open-file
  (kbd "m") 'dired-mark)
#+end_src

*** Treemacs
| Key Binding    | Command                   | Description                  |
|----------------+---------------------------+------------------------------|
| <LEADER> z t t | treemacs                  | Open Treemacs                |
| <LEADER> z t a | treemacs-create-file      | Create File                  |
| <LEADER> z t d | treemacs-create-dir       | Create Directory             |
| <LEADER> z t c | treemacs-select-directory | Select Directory in Treemacs |

#+begin_src emacs-lisp :tangle config.el
(defun zdmeyer-visit-treemacs() (interactive)
  (funcall 'treemacs)
  (funcall 'treemacs))

(map! :leader (:prefix ("z t" . "Treemacs")
               :desc "Toggle Treemacs" "t" #'treemacs
               :desc "Create File" "a" #'treemacs-create-file
               :desc "Create Directory" "d" #'treemacs-create-dir
               :desc "Change Directory" "c" #'treemacs-select-directory))

#+end_src

*** Term
| Key Binding    | Command             | Description                         |
|----------------+---------------------+-------------------------------------|
| <LEADER> z x v | multi-vterm-project | Opens a vterm for the given project |
| <LEADER> z x e | eshell              | Opens an eshell                     |

#+begin_src emacs-lisp :tangle config.el
(map! :leader
      (:prefix ("z x" . "Terminals")
       :desc "Open Multi-Vterm Project" "v" #'multi-vterm-project
       :desc "Open Eshell" "e" #'eshell))
#+end_src

** Documentation Comments
This section currently only has code for putting c++ comments for doxygen.

#+begin_src emacs-lisp :tangle config.el
;;
;; C Documentatin Comment Style
;;
(setq c-doc-comment-style '((c-mode . gtkdoc)
                            (c++-mode . doxygen)))
#+end_src

** Closing Comments
Preserve this crap.

#+begin_src emacs-lisp :tangle config.el
;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

#+end_src

* Packages.el
** Opening Comments
I like to preserve the comments in packages.el.

#+begin_src emacs-lisp :tangle packages.el
;; -*- no-byte-compile: t; -*-
;;; $DOOMDIR/packages.el

;; To install a package with Doom you must declare them here and run 'doom sync'
;; on the command line, then restart Emacs for the changes to take effect -- or
;; use 'M-x doom/reload'.


;; To install SOME-PACKAGE from MELPA, ELPA or emacsmirror:
;; (package! some-package)

;; To install a package directly from a remote git repo, you must specify a
;; `:recipe'. You'll find documentation on what `:recipe' accepts here:
;; https://github.com/radian-software/straight.el#the-recipe-format
;; (package! another-package
;;   :recipe (:host github :repo "username/repo"))

;; If the package you are trying to install does not contain a PACKAGENAME.el
;; file, or is located in a subdirectory of the repo, you'll need to specify
;; `:files' in the `:recipe':
;; (package! this-package
;;   :recipe (:host github :repo "username/repo"
;;            :files ("some-file.el" "src/lisp/*.el")))

;; If you'd like to disable a package included with Doom, you can do so here
;; with the `:disable' property:
;; (package! builtin-package :disable t)

;; You can override the recipe of a built in package without having to specify
;; all the properties for `:recipe'. These will inherit the rest of its recipe
;; from Doom or MELPA/ELPA/Emacsmirror:
;; (package! builtin-package :recipe (:nonrecursive t))
;; (package! builtin-package-2 :recipe (:repo "myfork/package"))

;; Specify a `:branch' to install a package from a particular branch or tag.
;; This is required for some packages whose default branch isn't 'master' (which
;; our package manager can't deal with; see radian-software/straight.el#279)
;; (package! builtin-package :recipe (:branch "develop"))

;; Use `:pin' to specify a particular commit to install.
;; (package! builtin-package :pin "1a2b3c4d5e")


;; Doom's packages are pinned to a specific commit and updated from release to
;; release. The `unpin!' macro allows you to unpin single packages...
;; (unpin! pinned-package)
;; ...or multiple packages
;; (unpin! pinned-package another-pinned-package)
;; ...Or *all* packages (NOT RECOMMENDED; will likely break things)
;; (unpin! t)
#+end_src

** Programming
#+begin_src emacs-lisp :tangle packages.el
(package! cov)
(package! coverlay)
(package! copilot
  :recipe (:host github :repo "copilot-emacs/copilot.el" :files ("*.el")))
(package! flycheck-clang-tidy
  :recipe (:host github :repo "Kazzira/flycheck-clang-tidy" :files ("*.el")))
#+end_src

** DOXYGEN

#+begin_src emacs-lisp :tangle packages.el
(package! highlight-doxygen)
#+end_src

** ORG MODE PACKAGES
Here are some of the org mode packages I use.
#+begin_src emacs-lisp :tangle packages.el
;; org-roam
(package! org-roam)
(package! org-bullets)
#+end_src

** Paredit
#+begin_src emacs-lisp :tangle packages.el
(package! paredit)
#+end_src

** Terminal
#+begin_src emacs-lisp :tangle packages.el
(package! multi-vterm)
#+end_src


doom/.config/doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
;; (setq user-full-name "John Doe"
;;       user-mail-address "john@doe.coM")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-symbol-font' -- for symbols
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!

(defun sa-find-org-file-recursively (&optional directory filext)
  "Return .org and .org_archive files recursively from DIRECTORY.
If FILEXT is provided, return files with extension FILEXT instead."
  (interactive "DDirectory: ")
  (let* (org-file-list
         (case-fold-search t)         ; filesystems are case sensitive
         (file-name-regex "^[^.#].*") ; exclude dot, autosave, and backupfiles
         (filext (or filext "org$\\\|org_archive"))
         (fileregex (format "%s\\.\\(%s$\\)" file-name-regex filext))
         (cur-dir-list (directory-files directory t file-name-regex)))
    ;; loop over directory listing
    (dolist (file-or-dir cur-dir-list org-file-list) ; returns org-file-list
      (cond
       ((file-regular-p file-or-dir)             ; regular files
        (if (string-match fileregex file-or-dir) ; org files
            (add-to-list 'org-file-list file-or-dir)))
       ((file-directory-p file-or-dir)
        (dolist (org-file (sa-find-org-file-recursively file-or-dir filext)
                          org-file-list) ; add files found to result
          (add-to-list 'org-file-list org-file)))))))

(setq org-directory "~/org/")
(setq org-agenda-files (append (sa-find-org-file-recursively "~/org/")))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'deeper-blue)

;; Set the font sizes for the org levels.
(custom-set-faces
 '(org-level-1 ((t (:inherit outline-1 :height 1.6))))
 '(org-level-2 ((t (:inherit outline-2 :height 1.3))))
 '(org-level-3 ((t (:inherit outline-3 :height 1.21))))
 '(org-level-4 ((t (:inherit outline-4 :height 1.15))))
 '(org-level-5 ((t (:inherit outline-5 :height 1.09))))
)

(setq org-hide-emphasis-markers t)

(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

;; This determines the style of line numbers in effect. If set to `nil', line
; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)

(require 'dap-gdb-lldb)

;;
;; C++ Development
;;
(with-eval-after-load 'lsp-clangd
  (require 'dap-cpptools)
  (require 'dap-lldb))

;;
;; C++ file syntax highlighting
;;
(add-to-list 'auto-mode-alist '(".*\.inl" . cpp-mode))
(add-to-list 'auto-mode-alist '(".*\.tpp" . cpp-mode))

;;
;; Dired Keybindings
;;
(map! :leader
      (:prefix ("d" . "dired")
       :desc "Open dired" "d" #'dired
       :desc "Dired jump to current" "j" #'dired-jump)
      (:after dired
        (:map dired-mode-map
         :desc "Peep-dired image previews" "d p" #'peep-dired
         :desc "Dired view file" "d v" #'dired-view-file)))

(evil-define-key 'normal dired-mode-map
  (kbd "M-RET") 'dired-display-file
  (kbd "h") 'dired-up-directory
  (kbd "l") 'dired-open-file
  (kbd "m") 'dired-mark)

(defun zdmeyer-visit-treemacs() (interactive)
  (funcall 'treemacs)
  (funcall 'treemacs))

(map! :leader (:prefix ("z t" . "Treemacs")
               :desc "Toggle Treemacs" "t" #'treemacs
               :desc "Create File" "a" #'treemacs-create-file
               :desc "Create Directory" "d" #'treemacs-create-dir
               :desc "Change Directory" "c" #'treemacs-select-directory))

(map! :leader
      (:prefix ("z x" . "Terminals")
       :desc "Open Multi-Vterm Project" "v" #'multi-vterm-project
       :desc "Open Eshell" "e" #'eshell))

;;
;; C Documentatin Comment Style
;;
(setq c-doc-comment-style '((c-mode . gtkdoc)
                            (c++-mode . doxygen)))

;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.
